### DISCUSSION TITLE
`render` function not working in nextjs 14

### DISCUSSION THREAD

**Discussion by mamlzy** [resend/react-email#1118] [Q&A] [answered]
I'm using `next` version 14.0.4
render function not working

here is the code:

<img width="323" alt="image" src="https://github.com/resend/react-email/assets/63343105/82332949-e7dd-4e63-87b4-b760ef838d6b">

the error:
<img width="725" alt="image" src="https://github.com/resend/react-email/assets/63343105/73d6fe33-9177-470b-8499-3095f01236b8">

is anybody facing the same issue like me?

**Comment by gabrielmfern** [comment]
Afaik this error is thrown due to nextjs overriding the usual implementation for renderToStaticMarkup. What I think is happening is that renderToStaticMarkup and other sync methods like this one are going to be deprecated and Next is already ready for that. 
The way to fix it is to not use the React's sync render method, and use the renderAsync one instead. 

I think we will just have `renderAsync` as `render` to avoid these issues. The latest version for `@react-email/render` should have a `renderAsync` that works in your situation. If you are running on the edge you might have issues with the latest one though and, if I remember correctly, they have already been fixed on the latest canary for `@react-email/render`.

**Reply by mamlzy** [reply]
yes i tried using `renderAsync` it is working fine, but `renderAsync` only works in RSC right? what i want to do is something like this:

<img width="349" alt="image" src="https://github.com/resend/react-email/assets/63343105/85d06d4e-0538-4bfd-ac71-e60d5d0cedbf">

but i want to rerender after some states changes. i did this is for make the "Email Template" updates based on what user typing from inputs.
here is the app:

<img width="1009" alt="image" src="https://github.com/resend/react-email/assets/63343105/1b4ec61e-7d47-4d71-acd8-d40e820bbf56">

can you give me the best practice for this? thanks!

**Reply by mamlzy** [reply]
i try to downgrade `nextjs` to `13.5.x` and the `render` function working fine. But i think downgrading is not the best solution🙁

**Reply by gabrielmfern** [reply]
`renderAsync` should work in other environments too, has it failed somewhere else for you?
The thing is that you are going to have to treat it like you treat fetch calls if you aren't using RSCs. Maybe on a useEffect and setting state to then render.

**Reply by gabrielmfern** [reply]
You could also directly render the email on your React component if you can't use `@react-email/render` there

**Comment by mamlzy** [comment]
> You could also directly render the email on your React component if you can't use `@react-email/render` there

it will show the same error if i directly render the Email Component in the `page.tsx`, that's why i am coverting the Email Template to `html` using `render` and pass it to the `srcDoc` `iframe`'s props

**Reply by gabrielmfern** [reply, marked as answer]
how about with something like:

```tsx
const [markup, setMarkup] = useState<string>();

useEffect(() => {
  (async () => {
    setMarkup(await renderAsync(<Customized customerName={name} />)
  })();
}, [debouncedName]);
```

---

TLDR: Use this alongside upgrading `@react-email/render` to canary that improved the renderAsync function and import from it to make sure it uses the proper version.

**Reply by mamlzy** [reply]
still showing the same erorr in nextjs 14😕, but working fine in next `13.5.x`. But @gabrielmfern have you tried this code in next 14? is it works in yours? please, let me know..

the code:
<img width="476" alt="image" src="https://github.com/resend/react-email/assets/63343105/05feedb5-4db1-4db9-aa54-4c94084966fc">

the error:
<img width="1270" alt="image" src="https://github.com/resend/react-email/assets/63343105/78d22dca-1bba-4690-90de-a6fb010c552d">

**Reply by gabrielmfern** [reply]
Can you try this out on the canary for render (`@react-email/render@canary`)? Also remember to import from `@react-email/render` instead of from components.

**Reply by mamlzy** [reply]
oh wait, after upgrading `@react-email/components` to `canary` version, your code is works in nextjs 14! thank you so much, but tell me is it good for performance that await-ing a promise inside useEffect, which is runs for every dependency array changes?

**Reply by mamlzy** [reply]
is importing from `@react-email/render` is better instead of `@react-email/components`?

**Reply by gabrielmfern** [reply]
> oh wait, after upgrading `@react-email/components` to `canary` version, your code is works in nextjs 14! thank you so much, but tell me is it good for performance that await-ing a promise inside useEffect, which is runs for every dependency array changes?

I don't think it would cause any performance issues but you could do it in prettier ways like using TanStack Query

**Reply by gabrielmfern** [reply]
> is importing from `@react-email/render` is better instead of `@react-email/components`?

I'd argue it's better to use from components for cleaner code in the end, but here it's important you import directly from `@react-email/render` because of the canary version. `@react-email/components` just re-exports it from a specific matching version of `@react-email/render` so it is basically the same thing

**Reply by mamlzy** [reply]
thanks for all @gabrielmfern!
